src/libs/filterArrayByMatch.ts

```diff
@@ -0,0 +1,327 @@
+        const inOrderPercentage = matchingInOrderCharCount / stringToRank.length;
+        const ranking = rankings.MATCHES + inOrderPercentage * spreadPercentage;
+
+        return ranking as Ranking;
+    }
+
+    const firstIndex = findMatchingCharacter(stringToRank[0], testString, 0);
+
+    if (firstIndex < 0) {
+        return rankings.NO_MATCH;
+    }
+
+    charNumber = firstIndex;
+    for (let i = 1; i < stringToRank.length; i++) {
+        const matchChar = stringToRank[i];
+        charNumber = findMatchingCharacter(matchChar, testString, charNumber);
+        const found = charNumber > -1;
+
+        if (!found) {
+            return rankings.NO_MATCH;
+        }
+    }
+
+    const spread = charNumber - firstIndex;
+    return getRanking(spread);
+}
+
+/**
+ * Gives a rankings score based on how well the two strings match.
+ * @param testString - the string to test against
+ * @param stringToRank - the string to rank
+ * @returns the ranking for how well stringToRank matches testString
+ */
+function getMatchRanking(testString: string, stringToRank: string): Ranking {
+    // too long
+    if (stringToRank.length > testString.length) {
+        return rankings.NO_MATCH;
+    }
+
+    // case sensitive equals
+    if (testString === stringToRank) {
+        return rankings.CASE_SENSITIVE_EQUAL;
+    }
+
+    // Lower casing before further comparison
+    const lowercaseTestString = testString.toLowerCase();
+    const lowercaseStringToRank = stringToRank.toLowerCase();
+
+    // case insensitive equals
+    if (lowercaseTestString === lowercaseStringToRank) {
+        return rankings.EQUAL;
+    }
+
+    // starts with
+    if (lowercaseTestString.startsWith(lowercaseStringToRank)) {
+        return rankings.STARTS_WITH;
+    }
+
+    // word starts with
+    if (lowercaseTestString.includes(` ${lowercaseStringToRank}`)) {
+        return rankings.WORD_STARTS_WITH;
+    }
+
+    // contains
+    if (lowercaseTestString.includes(lowercaseStringToRank)) {
+        return rankings.CONTAINS;
+    }
+    if (lowercaseStringToRank.length === 1) {
+        return rankings.NO_MATCH;
+    }
+
+    // acronym
+    if (getAcronym(lowercaseTestString).includes(lowercaseStringToRank)) {
+        return rankings.ACRONYM;
+    }
+
+    // will return a number between rankings.MATCHES and rankings.MATCHES + 1 depending  on how close of a match it is.
+    return getClosenessRanking(lowercaseTestString, lowercaseStringToRank);
+}
+
+/**
+ * Gets the highest ranking for value for the given item based on its values for the given keys
+ * @param item - the item to rank
+ * @param keys - the keys to get values from the item for the ranking
+ * @param value - the value to rank against
+ * @param options - options to control the ranking
+ * @returns the highest ranking
+ */
+function getHighestRanking<T>(item: T, keys: ReadonlyArray<KeyOption<T>> | undefined, value: string, options: Options<T>): RankingInfo {
+    if (!keys) {
```

Comment:
in our case `keys` is always provided, so I think we can make it required and get rid of this base case rather than building things we don't use or intend to use